# Supplementary tables (excel) column name descriptions

# Clumping and finemapping:
meta_finemapping_colname_descriptions <- c("LOCUS" = "locus number from clumping analysis",
                                           "CHROM" = "chromosome",
                                           "POS" = "base position in hg37",
                                           "ID" = "variant identifier",
                                           "REF" = "reference allele",
                                           "ALT" = "alternate allele",
                                           "studies" = "number of studies",
                                           "BETA" = "beta estimate",
                                           "SE" = "standard error of beta estimate",
                                           "CHISQ" = "chi-squared statistic",
                                           "P" = "p-value",
                                           "Q" = "Q statistic",
                                           "QP" = "Q p-value",
                                           "INFO" = "imputation accuracy",
                                           "AFCAS" = "allele frequency in cases",
                                           "AFCON" = "allele frequency in controls",
                                           "NCAS" = "number of cases",
                                           "NCON" = "number of controls",
                                           "NEFF" = "effective sample size",
                                           "NTOT" = "total sample size",
                                           "start" = "start position of the locus",
                                           "end" =  "end position of the locus"
                                             #,
  #                                          "Locus" = , 
  #                                          "MarkerName" = ,
  #                                          "Chromosome" = ,
  #                                          "Position" = ,
  #                                          "EA" = , "NEA", 
  # "EAF", "Nsample", "Ncohort", "Effects", "beta_0", "se_0", "beta_1", 
  # "se_1", "beta_2", "se_2", "beta_3", "se_3", "chisq_association", 
  # "ndf_association", "P-value_association", "chisq_ancestry_het", 
  # "ndf_ancestry_het", "P-value_ancestry_het", "chisq_residual_het", 
  # "ndf_residual_het", "P-value_residual_het", "lnBF", "Comments", 
  # "CHR", "start", "end",
  # 
  # "CHR", "BP_START", "BP_END", "CS_ID", "SNP", "BP", "REF_ALLELE_AFR", 
  # "REF_ALLELE_SAS", "REF_ALLELE_EUR", "ALT_ALLELE_AFR", "ALT_ALLELE_SAS", 
  # "ALT_ALLELE_EUR", "REF_FRQ_AFR", "REF_FRQ_SAS", "REF_FRQ_EUR", 
  # "BETA_AFR", "BETA_SAS", "BETA_EUR", "SE_AFR", "SE_SAS", "SE_EUR", 
  # "-LOG10P_AFR", "-LOG10P_SAS", "-LOG10P_EUR", "CS_PIP", "OVRL_PIP",
  # 
  # "CHR", "BP_START", "BP_END", "CS_ID", "CS_LENGTH", "CS_PURITY", 
  # "MAX_PIP_SNP", "BP", "REF_ALLELE_AFR", "REF_ALLELE_SAS", "REF_ALLELE_EUR", 
  # "ALT_ALLELE_AFR", "ALT_ALLELE_SAS", "ALT_ALLELE_EUR", "REF_FRQ_AFR", 
  # "REF_FRQ_SAS", "REF_FRQ_EUR", "BETA_AFR", "BETA_SAS", "BETA_EUR", 
  # "SE_AFR", "SE_SAS", "SE_EUR", "-LOG10P_AFR", "-LOG10P_SAS", "-LOG10P_EUR", 
  # "MAX_PIP", "POST-HOC_PROB_POP_AFR", "POST-HOC_PROB_POP_SAS", 
  # "POST-HOC_PROB_POP_EUR"
)