{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Meta-analysis of antidepressant exposure GWAS"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Conduct GWAS meta-analysis using [Hail](https://hail.is). Import GWAS sumstats in [GWAS VCF](https://github.com/MRCIEU/gwas-vcf-specification) format, which can be read into Hail and understood like any other VCF file. See also [meta-analysis functions from the Pan UK Biobank](https://github.com/atgu/ukbb_pan_ancestry/blob/master/run_ma.py). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "SLF4J: No SLF4J providers were found.\n",
      "SLF4J: Defaulting to no-operation (NOP) logger implementation\n",
      "SLF4J: See https://www.slf4j.org/codes.html#noProviders for further details.\n",
      "SLF4J: Class path contains SLF4J bindings targeting slf4j-api versions 1.7.x or earlier.\n",
      "SLF4J: Ignoring binding found at [jar:file:/home/madams23/miniconda3/envs/hail/lib/python3.8/site-packages/pyspark/jars/log4j-slf4j-impl-2.17.2.jar!/org/slf4j/impl/StaticLoggerBinder.class]\n",
      "SLF4J: See https://www.slf4j.org/codes.html#ignoredBindings for an explanation.\n",
      "Setting default log level to \"WARN\".\n",
      "To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).\n",
      "Running on Apache Spark version 3.3.2\n",
      "SparkUI available at http://waddington:4040\n",
      "Welcome to\n",
      "     __  __     <>__\n",
      "    / /_/ /__  __/ /\n",
      "   / __  / _ `/ / /\n",
      "  /_/ /_/\\_,_/_/_/   version 0.2.119-ca0ff87b1687\n",
      "LOGGING: writing to /home/madams23/Projects/antidep-gwas/hail-20230630-1026-0.2.119-ca0ff87b1687.log\n"
     ]
    }
   ],
   "source": [
    "import hail as hl\n",
    "hl.init(local='local[8]')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import GWAS VCFs and convert to `MatrixTable`s."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-06-30 10:29:52.722 Hail: INFO: scanning VCF for sortedness...\n",
      "2023-06-30 10:30:08.744 Hail: INFO: Coerced sorted VCF - no additional import work to do\n",
      "2023-06-30 10:30:37.893 Hail: INFO: wrote matrix table with 19364085 rows and 1 column in 5 partitions to mt/FinnGen-EUR.mt\n",
      "2023-06-30 10:30:38.057 Hail: INFO: scanning VCF for sortedness...\n",
      "2023-06-30 10:30:59.118 Hail: INFO: Coerced prefix-sorted VCF, requiring additional sorting within data partitions on each query.\n",
      "2023-06-30 10:31:29.085 Hail: INFO: wrote matrix table with 16399430 rows and 1 column in 5 partitions to mt/GenScot-EUR.mt\n",
      "2023-06-30 10:31:29.241 Hail: INFO: scanning VCF for sortedness...\n",
      "2023-06-30 10:31:50.655 Hail: INFO: Coerced sorted VCF - no additional import work to do\n",
      "2023-06-30 10:32:12.098 Hail: INFO: wrote matrix table with 13421318 rows and 1 column in 4 partitions to mt/BBJ-EAS.mt\n"
     ]
    }
   ],
   "source": [
    "from os import path\n",
    "import os\n",
    "import glob\n",
    "\n",
    "if not os.path.exists(\"mt\"):\n",
    "    os.makedirs(\"mt\")\n",
    "\n",
    "for gwas_vcf in glob.glob(\"vcf/*.vcf.gz\"):\n",
    "    dataset = path.basename(gwas_vcf).split(os.extsep)[0]\n",
    "    gwas_mt = path.join(\"mt\", os.extsep.join([dataset, \"mt\"]))\n",
    "    if not path.exists(gwas_mt):\n",
    "        hl.import_vcf(gwas_vcf,  reference_genome=\"GRCh38\", force_bgz=True).write(gwas_mt, overwrite=True)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Open all the sumstats tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "mts = [hl.read_matrix_table(mt) for mt in glob.glob(\"mt/*.mt\")]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Merge the list of sumstats together."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "import functools\n",
    "\n",
    "gw = functools.reduce(lambda mt1, mt2: mt1.union_cols(mt2, row_join_type = \"outer\"), mts)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hail",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
