---
title: "Supplementary Material"
output: word_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(glue)
library(stringr)


# ------------------------
# Define functions -------
# Function that takes a full path file name of a png,
# a regular expression to match a part of that file name
# and a figure legend using glue syntax for inserting the matched string,
# ie. it must contain "{file_str}" to be replaced by the matched string
# legend_title is in bold and legend_text is normal font
# also takes a numeric value for fig_number and offset for fig_number
# ie. the printed supplementary figure number will be fig_number + offset

paste_figure <- function(file, regex, legend_title, legend_text, fig_number, offset=0) {
  # check fig_number and offset are numeric
  if(!is.numeric(fig_number) | !is.numeric(offset)) {
    stop("fig_number and offset must be numeric")
  }
  file_str <- str_extract(file, regex)
  
  # Paste to render as markdown
  cat("![](", file, ")\n", sep = "")
  cat("\n\n")
  cat(paste0("**", glue("Supplementary Figure {fig_number + offset}"), ". ",  glue(legend_title), "** ",legend_text))
  cat("\n\n")
}

# Define offset value for starting supplementary figures at 1
offset <- 0
```

# Methods

## mBAT-combo analysis

SNPs with effective sample sizes < 0.8 or with duplicate IDs were removed from the meta-analysed summary statistics. Default values for mBAT-combo were used as follows: the minimum proportion of the total variance in the LD matrix explained by the top principal components used in the mBAT test was 0.9 (--mBAT-svd-gamma 0.9), a gene region was defined as spanning 50 kb upstream and downstream of the gene’s untranslated region (--mBAT-wind 50), SNPs with allele frequency difference between GWAS summary and LD reference data > 0.2 were removed (--diff-freq 0.2) and the threshold for LD r-squared value for LD pruning was 0.9 (--fastBAT-ld-cutoff 0.9).

# Results

## GWAS meta-analysis

```{r fixed_upset, results="asis", echo=FALSE}
fig_num <- 1
fig <- here::here("scripts", "fixed_files", "figure-gfm", "overlaps-1.png")
cat("![](", fig, ")\n\n", sep = "")
cat(paste0("**", glue("Supplementary Figure {fig_num}"), ". Number of idependent genome wide significant SNPs in each fixed meta-analysis.** Plot shows overlap in significant SNPs between antidepressant phenotype and ancestries. Genome wide significance is defined as p-value < 5e-8."))
offset <- fig_num
```
```{r multi_pcs_N06A, results="asis", echo=FALSE}
fig_num <- offset + 1
cat("![](", fig, ")\n\n", sep = "")
cat(paste0("**", glue("Supplementary Figure {fig_num}"), ". PCA plot of ancestries in MR-MEGA meta-analysis N06A.** Plot shows the first two principal components of the ancestry of the samples in the MR-MEGA meta-analysis for the antidepressant phenotype N06A."))
offset <- fig_num
```

```{r multi_pcs_N06AA, results="asis", echo=FALSE}
fig_num <- offset + 1
cat("![](", fig, ")\n\n", sep = "")
cat(paste0("**", glue("Supplementary Figure {fig_num}"), ". PCA plot of ancestries in MR-MEGA meta-analysis N06AA.** Plot shows the first two principal components of the ancestry of the samples in the MR-MEGA meta-analysis for the antidepressant phenotype N06AA."))
offset <- fig_num
```

```{r multi_pcs_N06AB, results="asis", echo=FALSE}
fig_num <- offset + 1
cat("![](", fig, ")\n\n", sep = "")
cat(paste0("**", glue("Supplementary Figure {fig_num}"), ". PCA plot of ancestries in MR-MEGA meta-analysis N06AB.** Plot shows the first two principal components of the ancestry of the samples in the MR-MEGA meta-analysis for the antidepressant phenotype N06AB."))
offset <- fig_num
```


```{r fixed_manhattan_N06A, results="asis", echo=FALSE}
path <- here::here("results", "meta", "antidep-2501")
files <- list.files(path, ".+fixed-N06A.+manhattan\\.png", full.names = T)
files <- files[!grepl("het|N06AA|N06AB", files)] # Exclude het and other antidep phenotypes

for(i in 1:length(files) ){
  paste_figure(files[i], 
             regex = "([A-Za-z0-9]+)(?=\\.manhattan\\.png)",
             legend_title = "Manhatan plot for fixed meta-analysis for N06A ancestry {file_str}.",
             legend_text = "The Manhattan plot shows the association of SNPs with the antidepressant phenotype N06A. The x-axis represents the chromosomal position of the SNP, and the y-axis represents the -log10(p-value) of the association. The red line represents the genome-wide significance threshold (p-value = 5e-8). The blue line represents the suggestive significance threshold (p-value = 1e-5).",
             fig_number = i,
             offset = offset)
}

# Keep a count on how many supplementary figures we have for the offset variable
offset <- offset + length(files)
```

```{r fixed_qq_N06A, results="asis", echo=FALSE}
path <- here::here("results", "meta", "antidep-2501")
files <- list.files(path, ".+fixed-N06A.+qq\\.png", full.names = T)
files <- files[!grepl("het|N06AA|N06AB", files)] # Exclude het and other antidep phenotypes

for(i in 1:length(files) ){
  paste_figure(files[i], 
             regex = "([A-Za-z0-9]+)(?=\\.qq\\.png)",
             legend_title = "QQ plot for fixed meta-analysis for N06A ancestry {file_str}.",
             legend_text = "QQ plot for the association of SNPs with the antidepressant phenotype N06A. The x-axis represents the expected -log10(p-value) of the association, and the y-axis represents the observed -log10(p-value) of the association. The lambda statistic (a measure of inflated p-values) is also reported on the plot.",
             fig_number = i,
             offset = offset)
}

# Keep a count on how many supplementary figures we have for the offset variable
offset <- offset + length(files)
```

```{r fixed_manhattan_N06AA, results="asis", echo=FALSE}
path <- here::here("results", "meta", "antidep-2501")
files <- list.files(path, ".+fixed-N06AA.+manhattan\\.png", full.names = T)
files <- files[!grepl("het", files)] # Exclude het 

for(i in 1:length(files) ){
  paste_figure(files[i], 
             regex = "([A-Za-z0-9]+)(?=\\.manhattan\\.png)",
             legend_title = "Manhatan plot for fixed meta-analysis for N06AA ancestry {file_str}.",
             legend_text = "The Manhattan plot shows the association of SNPs with the antidepressant phenotype N06AA. The x-axis represents the chromosomal position of the SNP, and the y-axis represents the -log10(p-value) of the association. The red line represents the genome-wide significance threshold (p-value = 5e-8). The blue line represents the suggestive significance threshold (p-value = 1e-5).",
             fig_number = i,
             offset = offset)
}

# Keep a count on how many supplementary figures we have for the offset variable
offset <- offset + length(files)
```

```{r fixed_qq_N06AA, results="asis", echo=FALSE}
path <- here::here("results", "meta", "antidep-2501")
files <- list.files(path, ".+fixed-N06AA.+qq\\.png", full.names = T)
files <- files[!grepl("het", files)] # Exclude het 

for(i in 1:length(files) ){
  paste_figure(files[i], 
             regex = "([A-Za-z0-9]+)(?=\\.qq\\.png)",
             legend_title = "QQ plot for fixed meta-analysis for N06AA ancestry {file_str}.",
             legend_text = "QQ plot for the association of SNPs with the antidepressant phenotype N06AA. The x-axis represents the expected -log10(p-value) of the association, and the y-axis represents the observed -log10(p-value) of the association. The lambda statistic (a measure of inflated p-values) is also reported on the plot.",
             fig_number = i,
             offset = offset)
}

# Keep a count on how many supplementary figures we have for the offset variable
offset <- offset + length(files)
```

```{r fixed_manhattan_N06AB, results="asis", echo=FALSE}
path <- here::here("results", "meta", "antidep-2501")
files <- list.files(path, ".+fixed-N06AB.+manhattan\\.png", full.names = T)
files <- files[!grepl("het", files)] # Exclude het 

for(i in 1:length(files) ){
  paste_figure(files[i], 
             regex = "([A-Za-z0-9]+)(?=\\.manhattan\\.png)",
             legend_title = "Manhatan plot for fixed meta-analysis for N06AB ancestry {file_str}.",
             legend_text = "The Manhattan plot shows the association of SNPs with the antidepressant phenotype N06AB. The x-axis represents the chromosomal position of the SNP, and the y-axis represents the -log10(p-value) of the association. The red line represents the genome-wide significance threshold (p-value = 5e-8). The blue line represents the suggestive significance threshold (p-value = 1e-5).",
             fig_number = i,
             offset = offset)
}

# Keep a count on how many supplementary figures we have for the offset variable
offset <- offset + length(files)
```

```{r fixed_qq_N06AB, results="asis", echo=FALSE}
path <- here::here("results", "meta", "antidep-2501")
files <- list.files(path, ".+fixed-N06AB.+qq\\.png", full.names = T)
files <- files[!grepl("het", files)] # Exclude het 

for(i in 1:length(files) ){
  paste_figure(files[i], 
             regex = "([A-Za-z0-9]+)(?=\\.qq\\.png)",
             legend_title = "QQ plot for fixed meta-analysis for N06AB ancestry {file_str}.",
             legend_text = "QQ plot for the association of SNPs with the antidepressant phenotype N06AB. The x-axis represents the expected -log10(p-value) of the association, and the y-axis represents the observed -log10(p-value) of the association. The lambda statistic (a measure of inflated p-values) is also reported on the plot.",
             fig_number = i,
             offset = offset)
}

# Keep a count on how many supplementary figures we have for the offset variable
offset <- offset + length(files)
```

```{r mrmega_manhattan_N06A, results="asis", echo=FALSE}
path <- here::here("results", "meta", "antidep-2501")
files <- list.files(path, ".+mrmega-N06A.+manhattan\\.png", full.names = T)
files <- files[!grepl("het|N06AA|N06AB", files)] # Exclude het and other antidep phenotypes

for(i in 1:length(files) ){
  paste_figure(files[i], 
             regex = "",
             legend_title = "Manhatan plot for MR-MEGA meta-analysis for N06A.",
             legend_text = "The Manhattan plot shows the association of SNPs with the antidepressant phenotype N06A. The x-axis represents the chromosomal position of the SNP, and the y-axis represents the -log10(p-value) of the association. The red line represents the genome-wide significance threshold (p-value = 5e-8). The blue line represents the suggestive significance threshold (p-value = 1e-5).",
             fig_number = i,
             offset = offset)
}

# Keep a count on how many supplementary figures we have for the offset variable
offset <- offset + length(files)
```

```{r mrmega_qq_N06A, results="asis", echo=FALSE}
path <- here::here("results", "meta", "antidep-2501")
files <- list.files(path, ".+mrmega-N06A.+qq\\.png", full.names = T)
files <- files[!grepl("het|N06AA|N06AB", files)] # Exclude het and other antidep phenotypes

for(i in 1:length(files) ){
  paste_figure(files[i], 
             regex = "",
             legend_title = "QQ plot for MR-MEGA meta-analysis for N06A.",
             legend_text = "QQ plot for the association of SNPs with the antidepressant phenotype N06A. The x-axis represents the expected -log10(p-value) of the association, and the y-axis represents the observed -log10(p-value) of the association. The lambda statistic (a measure of inflated p-values) is also reported on the plot.",
             fig_number = i,
             offset = offset)
}

# Keep a count on how many supplementary figures we have for the offset variable
offset <- offset + length(files)
```

```{r mrmega_manhattan_N06AA, results="asis", echo=FALSE}
path <- here::here("results", "meta", "antidep-2501")
files <- list.files(path, ".+mrmega-N06AA.+manhattan\\.png", full.names = T)
files <- files[!grepl("het", files)] # Exclude het and other antidep phenotypes

for(i in 1:length(files) ){
  paste_figure(files[i], 
             regex = "",
             legend_title = "Manhatan plot for MR-MEGA meta-analysis for N06AA.",
             legend_text = "The Manhattan plot shows the association of SNPs with the antidepressant phenotype N06AA. The x-axis represents the chromosomal position of the SNP, and the y-axis represents the -log10(p-value) of the association. The red line represents the genome-wide significance threshold (p-value = 5e-8). The blue line represents the suggestive significance threshold (p-value = 1e-5).",
             fig_number = i,
             offset = offset)
}

# Keep a count on how many supplementary figures we have for the offset variable
offset <- offset + length(files)
```

```{r mrmega_qq_N06AA, results="asis", echo=FALSE}
path <- here::here("results", "meta", "antidep-2501")
files <- list.files(path, ".+mrmega-N06AA.+qq\\.png", full.names = T)
files <- files[!grepl("het", files)] # Exclude het and other antidep phenotypes

for(i in 1:length(files) ){
  paste_figure(files[i], 
             regex = "",
             legend_title = "QQ plot for MR-MEGA meta-analysis for N06AA.",
             legend_text = "QQ plot for the association of SNPs with the antidepressant phenotype N06AA. The x-axis represents the expected -log10(p-value) of the association, and the y-axis represents the observed -log10(p-value) of the association. The lambda statistic (a measure of inflated p-values) is also reported on the plot.",
             fig_number = i,
             offset = offset)
}

# Keep a count on how many supplementary figures we have for the offset variable
offset <- offset + length(files)
```

```{r mrmega_manhattan_N06AB, results="asis", echo=FALSE}
path <- here::here("results", "meta", "antidep-2501")
files <- list.files(path, ".+mrmega-N06AB.+manhattan\\.png", full.names = T)
files <- files[!grepl("het", files)] # Exclude het and other antidep phenotypes

for(i in 1:length(files) ){
  paste_figure(files[i], 
             regex = "",
             legend_title = "Manhatan plot for MR-MEGA meta-analysis for N06AB.",
             legend_text = "The Manhattan plot shows the association of SNPs with the antidepressant phenotype N06AB. The x-axis represents the chromosomal position of the SNP, and the y-axis represents the -log10(p-value) of the association. The red line represents the genome-wide significance threshold (p-value = 5e-8). The blue line represents the suggestive significance threshold (p-value = 1e-5).",
             fig_number = i,
             offset = offset)
}

# Keep a count on how many supplementary figures we have for the offset variable
offset <- offset + length(files)
```

```{r mrmega_qq_N06AB, results="asis", echo=FALSE}
path <- here::here("results", "meta", "antidep-2501")
files <- list.files(path, ".+mrmega-N06AB.+qq\\.png", full.names = T)
files <- files[!grepl("het", files)] # Exclude het and other antidep phenotypes

for(i in 1:length(files) ){
  paste_figure(files[i], 
             regex = "",
             legend_title = "QQ plot for MR-MEGA meta-analysis for N06AB.",
             legend_text = "QQ plot for the association of SNPs with the antidepressant phenotype N06AB. The x-axis represents the expected -log10(p-value) of the association, and the y-axis represents the observed -log10(p-value) of the association. The lambda statistic (a measure of inflated p-values) is also reported on the plot.",
             fig_number = i,
             offset = offset)
}

# Keep a count on how many supplementary figures we have for the offset variable
offset <- offset + length(files)
```

