{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Antidepressant GWAS Phenotypes and Covariates\n",
    "\n",
    "Datasets were assembled in the AllofUs Workspace and then exported as queries."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "── \u001b[1mAttaching core tidyverse packages\u001b[22m ──────────────────────── tidyverse 2.0.0 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mdplyr    \u001b[39m 1.1.4     \u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.5\n",
      "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0     \u001b[32m✔\u001b[39m \u001b[34mstringr  \u001b[39m 1.5.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2  \u001b[39m 3.5.1     \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.2.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.3     \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.0.2     \n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n"
     ]
    }
   ],
   "source": [
    "library(tidyverse)\n",
    "library(bigrquery)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Queries"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get workspace variables and copy the queries to the local workspace."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "bucket <- Sys.getenv(\"WORKSPACE_BUCKET\")\n",
    "dataset <- Sys.getenv(\"WORKSPACE_CDR\")\n",
    "billing <- Sys.getenv(\"GOOGLE_PROJECT\")\n",
    "owner <- Sys.getenv(\"OWNER_EMAIL\")\n",
    "system(str_glue(\"gsutil -m cp {bucket}/queries/*.sql .\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [],
      "text/latex": [],
      "text/markdown": [],
      "text/plain": [
       "named character(0)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sql_paths <- list.files(\".\", \"*.sql\")\n",
    "names(sql_paths) <- sapply(str_split(sql_paths, \"\\\\.\"), first)\n",
    "sql_paths"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read SQL query from file, construct CSV output path and URI, perform query, and write out dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "for(query_name in names(sql_paths)) {\n",
    "    query_sql <- read_file(sql_paths[query_name])\n",
    "    query_csv <- str_glue(\"{query_name}.csv\")\n",
    "    bucket_query_csv_uri <- str_glue(\"{bucket}/bq_exports/{owner}/{query_csv}\")\n",
    "    query_tb <- bq_dataset_query(dataset, query = query_sql, billing = billing)\n",
    "    bq_table_save(query_tb, destination_uris = bucket_query_csv_uri, destination_format = \"CSV\")\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Copy saved CSV files into local workspace."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "system(str_glue(\"gsutil -m cp {bucket}/bq_exports/{owner}/*.csv .\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'genotyped-atc-n06a.csv'</li><li>'genotyped-atc-n06aa.csv'</li><li>'genotyped-atc-n06ab.csv'</li><li>'genotyped-ehr.csv'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'genotyped-atc-n06a.csv'\n",
       "\\item 'genotyped-atc-n06aa.csv'\n",
       "\\item 'genotyped-atc-n06ab.csv'\n",
       "\\item 'genotyped-ehr.csv'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'genotyped-atc-n06a.csv'\n",
       "2. 'genotyped-atc-n06aa.csv'\n",
       "3. 'genotyped-atc-n06ab.csv'\n",
       "4. 'genotyped-ehr.csv'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"genotyped-atc-n06a.csv\"  \"genotyped-atc-n06aa.csv\"\n",
       "[3] \"genotyped-atc-n06ab.csv\" \"genotyped-ehr.csv\"      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "list.files(\".\", \"*.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read in datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1mRows: \u001b[22m\u001b[34m206173\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m3\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \",\"\n",
      "\u001b[31mchr\u001b[39m (2): gender, date_of_birth\n",
      "\u001b[32mdbl\u001b[39m (1): person_id\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m2349104\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m6\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \",\"\n",
      "\u001b[31mchr\u001b[39m (3): standard_concept_name, standard_vocabulary, drug_exposure_start_dat...\n",
      "\u001b[32mdbl\u001b[39m (3): person_id, drug_concept_id, standard_concept_code\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m228304\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m6\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \",\"\n",
      "\u001b[31mchr\u001b[39m (3): standard_concept_name, standard_vocabulary, drug_exposure_start_dat...\n",
      "\u001b[32mdbl\u001b[39m (3): person_id, drug_concept_id, standard_concept_code\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m960626\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m6\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \",\"\n",
      "\u001b[31mchr\u001b[39m (3): standard_concept_name, standard_vocabulary, drug_exposure_start_dat...\n",
      "\u001b[32mdbl\u001b[39m (3): person_id, drug_concept_id, standard_concept_code\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n"
     ]
    }
   ],
   "source": [
    "genotyped_ehr <- read_csv(\"genotyped-ehr.csv\")\n",
    "atc_n06a <- read_csv(\"genotyped-atc-n06a.csv\")\n",
    "atc_n06aa <- read_csv(\"genotyped-atc-n06aa.csv\")\n",
    "atc_n06ab <- read_csv(\"genotyped-atc-n06ab.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Covariates\n",
    "\n",
    "Get provided ancestry cluster PCs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "system(str_glue(\"gsutil -m -u {billing} cp gs://fc-aou-datasets-controlled/v7/wgs/short_read/snpindel/aux/ancestry/ancestry_preds.tsv .\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1mRows: \u001b[22m\u001b[34m245394\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m5\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (4): ancestry_pred, probabilities, pca_features, ancestry_pred_other\n",
      "\u001b[32mdbl\u001b[39m (1): research_id\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n"
     ]
    }
   ],
   "source": [
    "ancestry_preds <- read_tsv(\"ancestry_preds.tsv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Parse out PCs from `pca_features` column (comma separated in brackets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "pcs_aux <- ancestry_preds |>\n",
    "    transmute(research_id, pca_features = str_sub(pca_features, start = 2, end = -2)) |>\n",
    "    separate_longer_delim(pca_features, delim = \",\") |>\n",
    "    transmute(research_id,\n",
    "              pc_value = as.numeric(pca_features)) |>\n",
    "    group_by(research_id) |>\n",
    "    mutate(PC = row_number()) |>\n",
    "    pivot_wider(names_prefix = \"PC\", names_from = PC, values_from = pc_value)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get PCs calculated from within each cluster (from `Ancestry cluster PCs.ipynb`):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "system(str_glue(\"gsutil -m cp {bucket}/pcs/*.sscore .\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "cols(\n",
      "  `#FID` = \u001b[32mcol_double()\u001b[39m,\n",
      "  IID = \u001b[32mcol_double()\u001b[39m,\n",
      "  ALLELE_CT = \u001b[32mcol_double()\u001b[39m,\n",
      "  NAMED_ALLELE_DOSAGE_SUM = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC1_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC2_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC3_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC4_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC5_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC6_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC7_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC8_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC9_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC10_AVG = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "cols(\n",
      "  `#FID` = \u001b[32mcol_double()\u001b[39m,\n",
      "  IID = \u001b[32mcol_double()\u001b[39m,\n",
      "  ALLELE_CT = \u001b[32mcol_double()\u001b[39m,\n",
      "  NAMED_ALLELE_DOSAGE_SUM = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC1_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC2_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC3_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC4_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC5_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC6_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC7_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC8_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC9_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC10_AVG = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "cols(\n",
      "  `#FID` = \u001b[32mcol_double()\u001b[39m,\n",
      "  IID = \u001b[32mcol_double()\u001b[39m,\n",
      "  ALLELE_CT = \u001b[32mcol_double()\u001b[39m,\n",
      "  NAMED_ALLELE_DOSAGE_SUM = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC1_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC2_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC3_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC4_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC5_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC6_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC7_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC8_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC9_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC10_AVG = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "cols(\n",
      "  `#FID` = \u001b[32mcol_double()\u001b[39m,\n",
      "  IID = \u001b[32mcol_double()\u001b[39m,\n",
      "  ALLELE_CT = \u001b[32mcol_double()\u001b[39m,\n",
      "  NAMED_ALLELE_DOSAGE_SUM = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC1_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC2_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC3_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC4_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC5_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC6_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC7_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC8_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC9_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC10_AVG = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "cols(\n",
      "  `#FID` = \u001b[32mcol_double()\u001b[39m,\n",
      "  IID = \u001b[32mcol_double()\u001b[39m,\n",
      "  ALLELE_CT = \u001b[32mcol_double()\u001b[39m,\n",
      "  NAMED_ALLELE_DOSAGE_SUM = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC1_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC2_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC3_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC4_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC5_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC6_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC7_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC8_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC9_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC10_AVG = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "cols(\n",
      "  `#FID` = \u001b[32mcol_double()\u001b[39m,\n",
      "  IID = \u001b[32mcol_double()\u001b[39m,\n",
      "  ALLELE_CT = \u001b[32mcol_double()\u001b[39m,\n",
      "  NAMED_ALLELE_DOSAGE_SUM = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC1_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC2_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC3_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC4_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC5_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC6_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC7_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC8_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC9_AVG = \u001b[32mcol_double()\u001b[39m,\n",
      "  PC10_AVG = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sscore_paths <- list.files(\".\", \"*.sscore\",)\n",
    "sscores <- bind_rows(lapply(sscore_paths, read_table))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "covariates_aux <- genotyped_ehr |>\n",
    "    inner_join(pcs_aux, by = c(\"person_id\" = \"research_id\")) |>\n",
    "    select(FID = person_id, IID = person_id,\n",
    "              gender, starts_with(\"PC\")) |>\n",
    "    mutate(gender = if_else(gender %in% c(\"Female\", \"Male\"), true = gender, false = \"Gender\")) |>\n",
    "    mutate(across(PC1:PC10, ~ scale(.x)[,1]))\n",
    "write_tsv(covariates_aux, \"atc_antidep_aux.covar\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "covariates <- genotyped_ehr |>\n",
    "    inner_join(sscores, by = c(\"person_id\" = \"IID\")) |>\n",
    "    select(FID = person_id, IID = person_id,\n",
    "              gender, starts_with(\"PC\")) |>\n",
    "    mutate(gender = if_else(gender %in% c(\"Female\", \"Male\"), true = gender, false = \"Gender\")) |>\n",
    "    mutate(across(PC1_AVG:PC10_AVG, ~ scale(.x)[,1]))\n",
    "write_tsv(covariates, \"atc_antidep.covar\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "206173"
      ],
      "text/latex": [
       "206173"
      ],
      "text/markdown": [
       "206173"
      ],
      "text/plain": [
       "[1] 206173"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(covariates_aux)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Phenotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotypes <- genotyped_ehr |>\n",
    "    transmute(FID = person_id, IID = person_id,\n",
    "             N06A = if_else(person_id %in% pull(atc_n06a, person_id), true = 1, false = 0),\n",
    "             N06AA = if_else(person_id %in% pull(atc_n06aa, person_id), true = 1, false = 0),\n",
    "             N06AB = if_else(person_id %in% pull(atc_n06ab, person_id), true = 1, false = 0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 5 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>N06A</th><th scope=col>N06AA</th><th scope=col>N06AB</th><th scope=col>n</th></tr>\n",
       "\t<tr><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>0</td><td>0</td><td>0</td><td>139379</td></tr>\n",
       "\t<tr><td>1</td><td>0</td><td>0</td><td> 16130</td></tr>\n",
       "\t<tr><td>1</td><td>0</td><td>1</td><td> 33752</td></tr>\n",
       "\t<tr><td>1</td><td>1</td><td>0</td><td>  8024</td></tr>\n",
       "\t<tr><td>1</td><td>1</td><td>1</td><td>  8888</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 5 × 4\n",
       "\\begin{tabular}{llll}\n",
       " N06A & N06AA & N06AB & n\\\\\n",
       " <dbl> & <dbl> & <dbl> & <int>\\\\\n",
       "\\hline\n",
       "\t 0 & 0 & 0 & 139379\\\\\n",
       "\t 1 & 0 & 0 &  16130\\\\\n",
       "\t 1 & 0 & 1 &  33752\\\\\n",
       "\t 1 & 1 & 0 &   8024\\\\\n",
       "\t 1 & 1 & 1 &   8888\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 5 × 4\n",
       "\n",
       "| N06A &lt;dbl&gt; | N06AA &lt;dbl&gt; | N06AB &lt;dbl&gt; | n &lt;int&gt; |\n",
       "|---|---|---|---|\n",
       "| 0 | 0 | 0 | 139379 |\n",
       "| 1 | 0 | 0 |  16130 |\n",
       "| 1 | 0 | 1 |  33752 |\n",
       "| 1 | 1 | 0 |   8024 |\n",
       "| 1 | 1 | 1 |   8888 |\n",
       "\n"
      ],
      "text/plain": [
       "  N06A N06AA N06AB n     \n",
       "1 0    0     0     139379\n",
       "2 1    0     0      16130\n",
       "3 1    0     1      33752\n",
       "4 1    1     0       8024\n",
       "5 1    1     1       8888"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "phenotypes |> count(N06A, N06AA, N06AB)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "write_tsv(phenotypes, \"atc_antidep.pheno\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "system(str_glue(\"gsutil -m cp atc_antidep.* {bucket}/inputs/\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.0"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
