---
title: Antidepressant exposure GWAS fixed-effects meta-analysis
output: github_document
---

```{r lib}
library(dplyr)
library(here)
library(readr)
library(stringr)
library(tidyr)
library(topr)
library(UpSetR)
library(plyranges)
library(ggplot2)
```


## GWAS results

### Sumstats

```{r sumstats, cache = TRUE, cache.lazy = FALSE}
# Use here to create paths relative to the top-level directory
# Specify which meta-analysis version to plot
metaset <- "antidep-2501"
# list fixed effects sumstats (.gz) files
sumstats_paths <- list.files(
  here::here("results", "meta", metaset),
  str_c(metaset, "-fixed-.+\\.gz"),
  full.names = TRUE
)
# simply names for plotting
prefixes <- str_remove(basename(sumstats_paths), ".gz")
metas <- str_remove(prefixes, str_c(metaset, "-fixed-"))

names(sumstats_paths) <- metas

sumstats <- lapply(sumstats_paths, function(path) {
  read_tsv(path) |>
    select(CHROM, POS, P)
})
```

### Loci

Collate all clump files together and mark which meta-analysis they are from.
```{r loci}
clumps_paths <- list.files(
  here::here("results", "meta", metaset),
  str_c(metaset, "-fixed-.+\\.clumps\\.tsv"),
  full.names = TRUE
)
clump_prefixes <- str_remove(basename(clumps_paths), ".clumps.tsv")
names(clumps_paths) <- str_remove(clump_prefixes, str_c(metaset, "-fixed-"))

clumps <- lapply(clumps_paths, read_table, col_types = cols(REF = col_character(), ALT = col_character()))

clumps_table <- bind_rows(clumps, .id = "dataset")
```

Write out clump file and column descriptions.
```{r table}
table_basename <- str_glue("clumps_fixed_{metaset}.clumps.csv")
write_csv(clumps_table, here::here("manuscript", "tables", table_basename))

colname_descriptions <-
  c("dataset" = "meta-analysis dataset (phenotype-cluster)",
    "LOCUS" = "locus number from clumping analysis",
    "CHROM" = "chromosome",
    "POS" = "base position in hg38",
    "ID" = "variant identifier",
    "REF" = "reference allele",
    "ALT" = "alternate allele",
    "studies" = "number of studies",
    "BETA" = "beta estimate",
    "SE" = "standard error of beta estimate",
    "CHISQ" = "chi-squared statistic",
    "P" = "p-value",
    "Q" = "Q statistic",
    "QP" = "Q p-value",
    "INFO" = "imputation accuracy",
    "AFCAS" = "allele frequency in cases",
    "AFCON" = "allele frequency in controls",
    "NCAS" = "number of cases",
    "NCON" = "number of controls",
    "NEFF" = "effective sample size",
    "NTOT" = "total sample size",
    "start" = "start position of the locus",
    "end" =  "end position of the locus")

colname_descriptions_table <- tibble(column = names(colname_descriptions), description = colname_descriptions)

# check that all names match
if (any(names(clumps_table) != names(colname_descriptions))) {
  stop(str_glue("column names in {table_basename}.csv are not all in colname_descriptions"))
}

write_tsv(colname_descriptions_table, here::here("manuscript", "tables", str_glue("{table_basename}.cols")))
```

## Manhattan plot

```{r manhattan, fig.width=8, fig.height=4.5, dpi=218}
sumstats2 <- lapply(sumstats, function(ss) filter(ss, P <= 1e-2))
manhattan(sumstats2, legend_labels = names(sumstats), ntop = 6, sign_thresh = 5e-08, build = 38)
```

## Loci

Construct genomic ranges.

```{r ranges}

clumped_ranges_grs <- lapply(clumps, function(cr) {
  cr |>
    as_granges(seqnames = CHROM, start = start, end = end) |>
    set_genome_info(genome = "hg38")
})

```

Count number of loci

```{r nloci}
sapply(clumped_ranges_grs, length)
```

## Overlaps

```{r overlaps, fig.width=8, fig.height=12, dpi=218}

all_gr <- reduce_ranges(bind_ranges(clumped_ranges_grs))

hits_upset <- lapply(clumped_ranges_grs, function(gr) findOverlaps(all_gr, gr)@from)

upset(fromList(hits_upset), nsets = length(hits_upset), order.by = "freq", text.scale = 2)

```

## Regions

```{r chr11_11, fig.width=8, fig.height=4.5, dpi=218}

regionplot(
  sumstats["N06A-EUR"], legend_labels = names(sumstats["N06A-EUR"]),
  chr = 11, xmin = 112956144, xmax = 113578846,
  build = 38, show_overview = FALSE
)
```