---
title: Antidepressant exposure GWAS fixed-effects meta-analysis
output: github_document
---

```{r lib}
library(dplyr)
library(here)
library(readr)
library(stringr)
library(tidyr)
library(topr)
library(UpSetR)
library(plyranges)
library(ggplot2)
```


## GWAS results

### Sumstats

```{r sumstats, cache = TRUE, cache.lazy = FALSE}
# Use here to create paths relative to the top-level directory

meta_gzs <- list.files(here::here("meta"), "antidep-2501-fixed-.+\\.gz", full.names = TRUE)
metas <- str_remove(meta_gzs, ".gz")
names(metas) <- str_remove(basename(metas), ".gz")
           
sumstats_paths <- lapply(metas, function(.x) str_c(.x, "gz", sep = "."))

sumstats <- lapply(meta_gzs, function(path) {
    read_tsv(path) |>
        select(CHROM, POS, ID,
               REF, ALT, P, OR,
               AF = AFCON, NCAS, NCON, NEFF)
})
```

### Loci

```{r loci}
clumped_paths <- sumstats_paths <- lapply(metas, function(.x) str_c(.x, "clumped", sep = "."))

clumped <- lapply(clumped_paths[sapply(clumped_paths, file.exists)], read_table)

ranges_paths <- sumstats_paths <- lapply(metas, function(.x) str_c(.x, "clumped.ranges", sep = "."))

clumped_ranges <- lapply(ranges_paths[sapply(ranges_paths, file.exists)], read_table)
```

## Manhattan plot

```{r manhattan, fig.width=8, fig.height=4.5, dpi=218}
sumstats2 <- lapply(sumstats, function(ss) filter(ss, P <= 1e-2))
manhattan(sumstats2, legend_labels = names(sumstats), ntop = 6, sign_thresh = 5e-08, build = 38)
```

## Loci

Construct genomic ranges.

```{r ranges}

clumped_ranges_grs <- lapply(clumped_ranges, function(cr) {
    cr |>
    mutate(range = str_match(POS, "chr[0-9]+:([0-9]+)\\.\\.([0-9]+)")) |>
    transmute(seqnames = str_c("chr", CHR),
              start = as.numeric(range[,2]),
              end = as.numeric(range[,3]), P, SNP) |>
    filter(P <= 5e-8) |>
    as_granges() |>
    set_genome_info(genome = "hg38")
})

loci_ranges_grs <- lapply(clumped_ranges_grs, function(gr) {
    reduce_ranges(gr, SNPs = c(SNP), Ps = c(P))
})

```

Count number of loci

```{r nloci}
sapply(loci_ranges_grs, length)
```

## Overlaps

```{r overlaps}

all_gr <- reduce_ranges(bind_ranges(loci_ranges_grs))

hits_upset <- lapply(loci_ranges_grs, function(gr) findOverlaps(all_gr, gr)@from)

upset(fromList(hits_upset), nsets = length(hits_upset), order.by='freq', text.scale=2)

```

## Regions

```{r chr11_11, fig.width=8, fig.height=4.5, dpi=218}

regionplot(sumstats["N06A-EUR"], legend_labels = names(sumstats["N06A-EUR"]),
    chr = 11, xmin = 112956144, xmax = 113578846,
    build = 38, show_overview = FALSE)

```