---
title: Antidepressant exposure GWAS fixed-effects meta-analysis
output: github_document
---

```{r lib, messages = FALSE}
library(dplyr)
library(here)
library(readr)
library(stringr)
library(tidyr)
library(topr)
library(UpSetR)
library(plyranges)
library(ggplot2)
```


## GWAS results

### Sumstats

Load all sumstats and label with `CLUSTER-PHENO`. Only retain chromsome, position, and p-value for plotting. Remove SNPs in alt regions.

```{r sumstats, cache = TRUE, cache.lazy = FALSE, messages = FALSE}
# Use here to create paths relative to the top-level directory
# Specify which meta-analysis version to plot
metaset <- "antidep-2501"
# list fixed effects sumstats (.gz) files
sumstats_paths <- list.files(
  here::here("results", "meta", metaset),
  str_c(metaset, "-fixed-.+\\.gz"),
  full.names = TRUE
)
# simply names for plotting
prefixes <- str_remove(basename(sumstats_paths), ".gz")
metas <- str_remove(prefixes, str_c(metaset, "-fixed-"))

names(sumstats_paths) <- metas

sumstats <- lapply(sumstats_paths, function(path) {
  read_tsv(path) |>
    select(CHROM, POS, P) |>
    # only keep chromosomes chr1-chr22 and chrX
    # (removes chromsomes with extra info on the end indicating
    #  and alt or unplaced region)
    filter(str_detect(CHROM, pattern = "^chr[0-9X]+$"))
})
```

### Loci

Collate all clump files together and mark which meta-analysis they are from.
```{r loci}
clumps_paths <- list.files(
  here::here("results", "meta", metaset),
  str_c(metaset, "-fixed-.+\\.clumps\\.tsv"),
  full.names = TRUE
)
clump_prefixes <- str_remove(basename(clumps_paths), ".clumps.tsv")
names(clumps_paths) <- str_remove(clump_prefixes, str_c(metaset, "-fixed-"))

clumps <- lapply(clumps_paths, read_table, col_types = cols(REF = col_character(), ALT = col_character()))

clumps_table <- bind_rows(clumps, .id = "dataset")
```

Write out clump file and column descriptions.
```{r table}
table_basename <- str_glue("clumps_fixed_{metaset}.clumps.csv")
write_csv(clumps_table, here::here("manuscript", "tables", table_basename))

colname_descriptions <-
  c("dataset" = "meta-analysis dataset (phenotype-cluster)",
    "LOCUS" = "locus number from clumping analysis",
    "CHROM" = "chromosome",
    "POS" = "base position in hg38",
    "ID" = "variant identifier",
    "REF" = "reference allele",
    "ALT" = "alternate allele",
    "studies" = "number of studies",
    "BETA" = "beta estimate",
    "SE" = "standard error of beta estimate",
    "CHISQ" = "chi-squared statistic",
    "P" = "p-value",
    "Q" = "Q statistic",
    "QP" = "Q p-value",
    "INFO" = "imputation accuracy",
    "AFCAS" = "allele frequency in cases",
    "AFCON" = "allele frequency in controls",
    "NCAS" = "number of cases",
    "NCON" = "number of controls",
    "NEFF" = "effective sample size",
    "NTOT" = "total sample size",
    "start" = "start position of the locus",
    "end" =  "end position of the locus")

colname_descriptions_table <- tibble(column = names(colname_descriptions), description = colname_descriptions)

# check that all names match
if (any(names(clumps_table) != names(colname_descriptions))) {
  stop(str_glue("column names in {table_basename}.csv are not all in colname_descriptions"))
}

write_tsv(colname_descriptions_table, here::here("manuscript", "tables", str_glue("{table_basename}.cols")))
```

## Manhattan plots

Find smallest p-value to draw all Manhattan plots on the same scale.
```{r min_p}
min_p <- min(sapply(sumstats, function(sumstat) min(sumstat$P, na.rm = TRUE)))
```
Assign consistent color to each cluster
```{r clusters}
clusters <- unique(sapply(str_split(names(sumstats), pattern = "-"), last))
cluster_colors <- get_topr_colors()[seq_along(clusters)]
names(cluster_colors) <- clusters
```

Function to plot a single phenotype.
```{r manh_pheno}
manh_pheno <- function(pheno, sumstats, min_p, cluster_colors) {
  sumstats_pheno <- sumstats[which(str_detect(names(sumstats), str_glue("{pheno}-")))]
  clusters_pheno <- sapply(str_split(names(sumstats_pheno), pattern = "-"), last)
  manhattan(
    sumstats_pheno,
    legend_labels = clusters_pheno,
    color = cluster_colors[clusters_pheno],
    ntop = length(sumstats_pheno),
    sign_thresh = 5e-08,
    ymax = ceiling(-log10(min_p)),
    build = 38,
    chr_ticknames = c(1:22, "X")
  )
}
```

### N06A

```{r manhattan_n06a, fig.width=12, fig.height=6.75, dpi=218, cache = TRUE, cache.lazy = FALSE}
manh_pheno("N06A", sumstats, min_p, cluster_colors)
```

### N06AA

```{r manhattan_n06aa, fig.width=12, fig.height=6.75, dpi=218, cache = TRUE, cache.lazy = FALSE}
manh_pheno("N06AA", sumstats, min_p, cluster_colors)
```

### N06AB

```{r manhattan_n06ab, fig.width=12, fig.height=6.75, dpi=218, cache = TRUE, cache.lazy = FALSE}
manh_pheno("N06AB", sumstats, min_p, cluster_colors)
```


## Loci

Construct genomic ranges.

```{r ranges}

clumped_ranges_grs <- lapply(clumps, function(cr) {
  cr |>
    as_granges(seqnames = CHROM, start = start, end = end) |>
    set_genome_info(genome = "hg38")
})

```

Count number of loci

```{r nloci}
sapply(clumped_ranges_grs, length)
```

## Overlaps

```{r overlaps, fig.width=8, fig.height=12, dpi=218}

all_gr <- reduce_ranges(bind_ranges(clumped_ranges_grs))

hits_upset <- lapply(clumped_ranges_grs, function(gr) findOverlaps(all_gr, gr)@from)

upset(fromList(hits_upset), nsets = length(hits_upset), order.by = "freq", text.scale = 2)

```

## Regions

```{r chr11_11, fig.width=8, fig.height=4.5, dpi=218}

regionplot(
  sumstats["N06A-EUR"], legend_labels = names(sumstats["N06A-EUR"]),
  chr = 11, xmin = 112956144, xmax = 113578846,
  build = 38, show_overview = FALSE
)
```