---
title: Antidepressant exposure GWAS meta-analysis loci
output: github_document
---

```{r lib, messages = FALSE}
library(dplyr)
library(here)
library(readr)
library(stringr)
library(tidyr)
library(topr)
library(UpSetR)
library(plyranges)
library(ggplot2)
library(readxl)
library(easylift)
```

```{r metaset}
metaset <- "antidep-2501"
```

## Clumps


Collate all clump files together and mark which meta-analysis they are from.
```{r loci}
clumps_paths <- list.files(
  here::here("results", "meta", metaset),
  str_c(metaset, ".+\\.clumps\\.tsv"),
  full.names = TRUE
)
clump_prefixes <- str_remove(basename(clumps_paths), ".clumps.tsv")

pheno_cluster <- sapply(str_split(clump_prefixes, pattern = "-"), function(x) str_c(x[4], x[5], sep = "-"))
pheno_cluster <- str_replace(pheno_cluster, "DIV", "MRMEGA")

names(clumps_paths) <- pheno_cluster

clumps <- lapply(clumps_paths, read_table, col_types = cols(REF = col_character(), ALT = col_character()))
```

## Genomics ranges

Construct genomic ranges.

```{r ranges}

clumped_ranges_grs <- lapply(clumps, function(cr) {
  cr |>
    select(
      CHROM = matches("chrom"),
      start,
      end,
      ID = matches("^(ID|MarkerName)$")) |>
    mutate(
      CHROM = if_else(
        str_detect(CHROM, "chr"),
        true = as.character(CHROM),
        false = str_c("chr", CHROM)
      )
    ) |>
    as_granges(seqnames = CHROM, start = start, end = end) |>
    set_genome_info(genome = "hg38")
})
```

Count number of loci

```{r datasets}
all_datasets <- names(clumped_ranges_grs)
all_n06a <- str_subset(names(clumped_ranges_grs), "N06A-")
all_n06aa <- str_subset(names(clumped_ranges_grs), "N06AA-")
all_n06ab <- str_subset(names(clumped_ranges_grs), "N06AB-")
fixed_datasets <- str_subset(names(clumped_ranges_grs), "MEGA", negate = TRUE)
fixed_n06a_datasets <- str_subset(names(clumped_ranges_grs), "N06A-[^M]")
fixed_n06aa_datasets <- str_subset(names(clumped_ranges_grs), "N06AA-[^M]")
fixed_n06ab_datasets <- str_subset(names(clumped_ranges_grs), "N06AB-[^M]")
mega_datasets <- str_subset(names(clumped_ranges_grs), "MEGA")
mega_n06a_datasets <- str_subset(names(clumped_ranges_grs), "N06A-MRMEGA")
mega_n06aa_datasets <- str_subset(names(clumped_ranges_grs), "N06AA-MRMEGA")
mega_n06ab_datasets <- str_subset(names(clumped_ranges_grs), "N06AB-MRMEGA")

datasets_lists <- list(
  all = all_datasets,
  all_n06a = all_n06a,
  all_n06ab = all_n06aa,
  all_n06ab = all_n06ab,
  fixed = fixed_datasets,
  fixed_n06a = fixed_n06a_datasets,
  fixed_n06aa = fixed_n06aa_datasets,
  fixed_n06ab = fixed_n06ab_datasets,
  mega = mega_datasets,
  mega_n06a = mega_n06a_datasets,
  mega_n06aa = mega_n06aa_datasets,
  mega_n06ab = mega_n06ab_datasets
)
```

```{r nloci}
sapply(datasets_lists, function(x) sum(sapply(clumped_ranges_grs[x], length)))
```

## Overlaps

```{r overlaps, fig.width=8, fig.height=12, dpi=218}

all_gr <- reduce_ranges(bind_ranges(clumped_ranges_grs))

hits_upset <- lapply(clumped_ranges_grs, function(gr) findOverlaps(all_gr, gr)@from)

upset(fromList(hits_upset), nsets = length(hits_upset), order.by = "freq", text.scale = 2)
```

Number of unique loci total

```{r unique}
sapply(datasets_lists, function(x) length(reduce_ranges(bind_ranges(clumped_ranges_grs[x]))))
```

SNPs unique to MR-MEGA analyses

```{r nonoverlap_mega}
filter_by_non_overlaps(
  clumped_ranges_grs[['N06AB-MRMEGA']],
  clumped_ranges_grs[['N06A-EUR']]
)

filter_by_non_overlaps(
  clumped_ranges_grs[['N06AB-MRMEGA']],
  bind_ranges(
    clumped_ranges_grs[['N06A-EUR']],
    clumped_ranges_grs[['N06AB-EUR']]
  )
)
```


## MDD2025

Compare to PGC MDD
```{r mdd2025}
mdd2025_all <- read_excel(here::here("manuscript/reference/MDD_GWAS_Online_Results_(COJO).xlsx"), sheet = 2) |>
  mutate(P = as.numeric(P))

mdd2025_eur <- read_excel(here::here("manuscript/reference/MDD_GWAS_Online_Results_(COJO).xlsx"), sheet = 3)

mdd2025_list <- list(mdd2025_all = mdd2025_all, mdd2025_eur = mdd2025_eur)

mdd2025_hg19_grs <- lapply(mdd2025_list, function(cr) {
  cr |>
    transmute(
      CHROM = if_else(CHR == 23, true = "chrX", false = str_c("chr", CHR)),
      start = Locus.Left,
      end = Locus.Right,
      SNP
    ) |>
    as_granges(seqnames = CHROM, start = start, end = end) |>
    set_genome_info(genome = "hg19")
})

chain <- here::here("reference/hg19ToHg38.over.chain.gz")

mdd2025_grs <- lapply(mdd2025_hg19_grs, function(gr) easylift(gr, "hg38", chain))

mdd2025_gr <- reduce_ranges(bind_ranges(mdd2025_grs))
```

Overlaps
```{r}
length(filter_by_overlaps(all_gr, mdd2025_gr))
```